lib/handlers/router-entities/route-caching/dynamo-route-caching-provider.ts

```diff
@@ -0,0 +1,302 @@
+   */
+  ttl_minutes?: number
+}
+
+export class DynamoRouteCachingProvider extends IRouteCachingProvider {
+  private readonly ddbClient: DynamoDB.DocumentClient
+  private readonly tableName: string
+  private readonly ttl_minutes: number
+
+  constructor({ cachedRoutesTableName, ttl_minutes = 5 }: ConstructorParams) {
+    super()
+    this.ddbClient = new DynamoDB.DocumentClient()
+    this.tableName = cachedRoutesTableName
+    this.ttl_minutes = ttl_minutes
+  }
+
+  /**
+   * Implementation of the abstract method defined in `IRouteCachingProvider`
+   * Given a CachedRoutesStrategy (from CACHED_ROUTES_CONFIGURATION) we will find the BlocksToLive associated to the bucket.
+   *
+   * @param cachedRoutes
+   * @param amount
+   * @protected
+   */
+  protected async _getBlocksToLive(cachedRoutes: CachedRoutes, amount: CurrencyAmount<Currency>): Promise<number> {
+    const cachedRoutesStrategy = this.getCachedRoutesStrategyFromCachedRoutes(cachedRoutes)
+    const cachingParameters = cachedRoutesStrategy?.getCachingParameters(amount)
+
+    if (cachingParameters) {
+      return cachingParameters.blocksToLive
+    } else {
+      return 0
+    }
+  }
+
+  /**
+   * Implementation of the abstract method defined in `IRouteCachingProvider`
+   * Fetch the most recent entry from the DynamoDB table for that pair, tradeType, chainId, protocols and bucket
+   *
+   * @param chainId
+   * @param amount
+   * @param quoteToken
+   * @param tradeType
+   * @param protocols
+   * @protected
+   */
+  protected async _getCachedRoute(
+    chainId: ChainId,
+    amount: CurrencyAmount<Currency>,
+    quoteToken: Token,
+    tradeType: TradeType,
+    protocols: Protocol[]
+  ): Promise<CachedRoutes | undefined> {
+    const [tokenIn, tokenOut] = this.determineTokenInOut(amount, quoteToken, tradeType)
+    const cachedRoutesStrategy = this.getCachedRoutesStrategy(tokenIn, tokenOut, tradeType, chainId)
+    const cachingParameters = cachedRoutesStrategy?.getCachingParameters(amount)
+
+    if (cachingParameters) {
+      const partitionKey = `${tokenIn.address}/${tokenOut.address}/${tradeType}/${chainId}`
+      const partialSortKey = `${protocols}/${cachingParameters.bucket}/`
+
+      const queryParams = {
+        TableName: this.tableName,
+        // Since we don't know what's the latest block that we have in cache, we make a query with a partial sort key
+        KeyConditionExpression: '#pk = :pk and begins_with(#sk, :sk)',
+        ExpressionAttributeNames: {
+          '#pk': 'pairTradeTypeChainId',
+          '#sk': 'protocolsBucketBlockNumber',
+        },
+        ExpressionAttributeValues: {
+          ':pk': partitionKey,
+          ':sk': partialSortKey,
+        },
+        ScanIndexForward: false, // Reverse order to retrieve most recent item first
+        Limit: 1, // Only retrieve the most recent item
+      }
+
+      try {
+        log.info({ queryParams }, `[DynamoRouteCachingProvider] Attempting to get route from cache.`)
```

Comment:
Is this `log` imported from smart order router? Shouldn't we be using the one in routing-api?

Reply:
I think the `log` comes from smart-order-router and it is injected from the injected-sor, so it is the same log